File: src/crossalpha/observatory/canonical/hyperliquid.py
from __future__ import annotations

import gzip
import json
from pathlib import Path
from typing import Any

import pandas as pd

from crossalpha.domain.models import RawSnapshotManifest
from crossalpha.observatory.health import load_manifest


def _to_float(value: Any) -> float | None:
    if value is None or value == "":
        return None
    try:
        return float(value)
    except (TypeError, ValueError):
        return None


def parse_meta_and_asset_contexts(envelope: dict[str, Any], raw_record: RawSnapshotManifest) -> pd.DataFrame:
    payload = envelope.get("payload")
    if not isinstance(payload, list) or len(payload) != 2:
        raise ValueError("Hyperliquid metaAndAssetCtxs payload must be [meta, asset_contexts]")

    meta, contexts = payload
    if not isinstance(meta, dict) or not isinstance(contexts, list):
        raise ValueError("Hyperliquid metaAndAssetCtxs payload shape is invalid")

    universe = meta.get("universe")
    if not isinstance(universe, list):
        raise ValueError("Hyperliquid meta.universe is missing")
    if len(universe) != len(contexts):
        raise ValueError(f"Hyperliquid universe/context length mismatch: {len(universe)} != {len(contexts)}")

    observed_at = envelope.get("observed_at")
    known_at = envelope.get("known_at")
    rows: list[dict[str, Any]] = []
    for spec, ctx in zip(universe, contexts, strict=True):
        if not isinstance(spec, dict) or not isinstance(ctx, dict):
            raise ValueError("Hyperliquid universe/context item is not an object")
        impact = ctx.get("impactPxs")
        impact_bid = _to_float(impact[0]) if isinstance(impact, list) and len(impact) > 0 else None
        impact_ask = _to_float(impact[1]) if isinstance(impact, list) and len(impact) > 1 else None
        rows.append(
            {
                "observed_at": observed_at,
                "known_at": known_at,
                "asset": spec.get("name"),
                "sz_decimals": spec.get("szDecimals"),
                "max_leverage": spec.get("maxLeverage"),
                "only_isolated": spec.get("onlyIsolated"),
                "mark_price": _to_float(ctx.get("markPx")),
                "oracle_price": _to_float(ctx.get("oraclePx")),
                "mid_price": _to_float(ctx.get("midPx")),
                "prev_day_price": _to_float(ctx.get("prevDayPx")),
                "premium": _to_float(ctx.get("premium")),
                "funding_rate": _to_float(ctx.get("funding")),
                "open_interest": _to_float(ctx.get("openInterest")),
                "day_notional_volume": _to_float(ctx.get("dayNtlVlm")),
                "day_base_volume": _to_float(ctx.get("dayBaseVlm")),
                "impact_bid": impact_bid,
                "impact_ask": impact_ask,
                "raw_sha256": raw_record.sha256,
                "raw_path": raw_record.path,
            }
        )

    frame = pd.DataFrame(rows)
    if frame.empty:
        raise ValueError("Hyperliquid canonicalization produced zero assets")
    if frame["asset"].isna().any() or frame["asset"].duplicated().any():
        raise ValueError("Hyperliquid canonical asset names are missing or duplicated")
    return frame


def canonicalize_hyperliquid(data_root: Path) -> dict[str, int]:
    records, errors = load_manifest(data_root)
    if errors:
        raise ValueError(f"raw manifest contains errors: {errors}")

    selected = [
        record
        for record in records
        if record.source_id == "hyperliquid" and record.observation_type == "metaAndAssetCtxs"
    ]
    written = 0
    skipped = 0
    rows = 0

    for record in selected:
        observed = record.observed_at
        out_dir = (
            data_root
            / "canonical"
            / "hyperliquid"
            / "asset_contexts"
            / f"year={observed:%Y}"
            / f"month={observed:%m}"
            / f"day={observed:%d}"
        )
        out_dir.mkdir(parents=True, exist_ok=True)
        out_path = out_dir / f"{observed:%Y%m%dT%H%M%S.%fZ}_{record.sha256[:12]}.parquet"
        if out_path.exists():
            skipped += 1
            continue

        with gzip.open(record.path, "rt", encoding="utf-8") as fh:
            envelope = json.load(fh)
        frame = parse_meta_and_asset_contexts(envelope, record)
        frame.to_parquet(out_path, index=False)
        written += 1
        rows += len(frame)

    return {"snapshots": len(selected), "written": written, "skipped": skipped, "rows_written": rows}

File: crates/crossalpha-features/src/canonical/hyperliquid.rs
use anyhow::{Context, Result, bail};
use chrono::{DateTime, Utc};
use crossalpha_storage::{ObservationEnvelope, RawSnapshotManifest};
use serde::Serialize;
use serde_json::Value;
use std::collections::HashSet;

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct HyperliquidAssetContextRow {
    pub observed_at: DateTime<Utc>,
    pub known_at: DateTime<Utc>,
    pub asset: String,
    pub sz_decimals: Value,
    pub max_leverage: Value,
    pub only_isolated: Value,
    pub mark_price: Option<f64>,
    pub oracle_price: Option<f64>,
    pub mid_price: Option<f64>,
    pub prev_day_price: Option<f64>,
    pub premium: Option<f64>,
    pub funding_rate: Option<f64>,
    pub open_interest: Option<f64>,
    pub day_notional_volume: Option<f64>,
    pub day_base_volume: Option<f64>,
    pub impact_bid: Option<f64>,
    pub impact_ask: Option<f64>,
    pub raw_sha256: String,
    pub raw_path: String,
}

pub fn parse_meta_and_asset_contexts(
    envelope: &ObservationEnvelope,
    raw_record: &RawSnapshotManifest,
) -> Result<Vec<HyperliquidAssetContextRow>> {
    let payload = envelope
        .payload
        .as_array()
        .context("Hyperliquid metaAndAssetCtxs payload must be [meta, asset_contexts]")?;
    if payload.len() != 2 {
        bail!("Hyperliquid metaAndAssetCtxs payload must be [meta, asset_contexts]");
    }

    let meta = payload[0]
        .as_object()
        .context("Hyperliquid metaAndAssetCtxs payload shape is invalid")?;
    let contexts = payload[1]
        .as_array()
        .context("Hyperliquid metaAndAssetCtxs payload shape is invalid")?;
    let universe = meta
        .get("universe")
        .and_then(Value::as_array)
        .context("Hyperliquid meta.universe is missing")?;
    if universe.len() != contexts.len() {
        bail!(
            "Hyperliquid universe/context length mismatch: {} != {}",
            universe.len(),
            contexts.len()
        );
    }

    let mut rows = Vec::with_capacity(universe.len());
    let mut assets = HashSet::with_capacity(universe.len());
    for (spec_value, ctx_value) in universe.iter().zip(contexts.iter()) {
        let spec = spec_value
            .as_object()
            .context("Hyperliquid universe/context item is not an object")?;
        let ctx = ctx_value
            .as_object()
            .context("Hyperliquid universe/context item is not an object")?;
        let asset = spec
            .get("name")
            .and_then(Value::as_str)
            .context("Hyperliquid canonical asset name is missing")?
            .to_owned();
        if !assets.insert(asset.clone()) {
            bail!("Hyperliquid canonical asset names are missing or duplicated");
        }

        let impact = ctx.get("impactPxs").and_then(Value::as_array);
        let impact_bid = impact.and_then(|items| items.first()).and_then(to_float);
        let impact_ask = impact.and_then(|items| items.get(1)).and_then(to_float);

        rows.push(HyperliquidAssetContextRow {
            observed_at: envelope.observed_at,
            known_at: envelope.known_at,
            asset,
            sz_decimals: spec.get("szDecimals").cloned().unwrap_or(Value::Null),
            max_leverage: spec.get("maxLeverage").cloned().unwrap_or(Value::Null),
            only_isolated: spec.get("onlyIsolated").cloned().unwrap_or(Value::Null),
            mark_price: ctx.get("markPx").and_then(to_float),
            oracle_price: ctx.get("oraclePx").and_then(to_float),
            mid_price: ctx.get("midPx").and_then(to_float),
            prev_day_price: ctx.get("prevDayPx").and_then(to_float),
            premium: ctx.get("premium").and_then(to_float),
            funding_rate: ctx.get("funding").and_then(to_float),
            open_interest: ctx.get("openInterest").and_then(to_float),
            day_notional_volume: ctx.get("dayNtlVlm").and_then(to_float),
            day_base_volume: ctx.get("dayBaseVlm").and_then(to_float),
            impact_bid,
            impact_ask,
            raw_sha256: raw_record.sha256.clone(),
            raw_path: raw_record.path.clone(),
        });
    }

    if rows.is_empty() {
        bail!("Hyperliquid canonicalization produced zero assets");
    }
    Ok(rows)
}

fn to_float(value: &Value) -> Option<f64> {
    match value {
        Value::Number(number) => number.as_f64(),
        Value::String(text) if !text.is_empty() => text.parse::<f64>().ok(),
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::TimeZone;
    use serde_json::{Map, json};

    fn manifest() -> RawSnapshotManifest {
        RawSnapshotManifest {
            path: "/tmp/raw.json.gz".to_owned(),
            sha256: "abc123".to_owned(),
            bytes: 1,
            compressed_bytes: Some(1),
            observed_at: Utc.timestamp_opt(1_700_000_000, 0).unwrap(),
            source_id: "hyperliquid".to_owned(),
            observation_type: "metaAndAssetCtxs".to_owned(),
        }
    }

    fn envelope(payload: Value) -> ObservationEnvelope {
        let now = Utc.timestamp_opt(1_700_000_000, 0).unwrap();
        ObservationEnvelope {
            schema_version: 1,
            event_time: None,
            observed_at: now,
            known_at: now,
            source_type: "EXCHANGE".to_owned(),
            source_id: "hyperliquid".to_owned(),
            observation_type: "metaAndAssetCtxs".to_owned(),
            payload,
            metadata: Map::new(),
        }
    }

    #[test]
    fn parses_python_contract_fields() {
        let payload = json!([
            {"universe": [{"name": "BTC", "szDecimals": 5, "maxLeverage": 40}]},
            [{
                "markPx": "100.5",
                "oraclePx": "100",
                "midPx": "100.25",
                "prevDayPx": "99",
                "premium": "0.001",
                "funding": "0.0001",
                "openInterest": "12",
                "dayNtlVlm": "1234",
                "dayBaseVlm": "5",
                "impactPxs": ["100.1", "100.4"]
            }]
        ]);
        let rows = parse_meta_and_asset_contexts(&envelope(payload), &manifest()).unwrap();
        assert_eq!(rows.len(), 1);
        assert_eq!(rows[0].asset, "BTC");
        assert_eq!(rows[0].mark_price, Some(100.5));
        assert_eq!(rows[0].impact_bid, Some(100.1));
        assert_eq!(rows[0].impact_ask, Some(100.4));
    }

    #[test]
    fn rejects_universe_context_length_mismatch() {
        let payload = json!([{"universe": [{"name": "BTC"}]}, []]);
        assert!(parse_meta_and_asset_contexts(&envelope(payload), &manifest()).is_err());
    }
}
